jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,427 @@
+     * The entity will never be replaced.
+     * @param entity the entity to add
+     */
+    @Override
+    public void addEntity(Entity entity) {
+        LOGGER.debug("Adding/Updating the entity in ES");
+
+        //The field that has a not null value in the current Entity instance.
+        //Required, mainly for building a script for the update command.
+        String fieldToAdd = null;
+        String fieldValueToAdd = "";
+        try {
+            XContentBuilder builder = jsonBuilder()
+                    .startObject();
+
+            //Currently ignoring Graph field based indexing
+//            if (docDef.getGraphField() != null) {
+//                builder = builder.field(docDef.getGraphField(), entity.getGraph());
+//            }
+
+            for(String field: docDef.fields()) {
+                if(entity.get(field) != null) {
+                    if(entity.getLanguage() != null && !entity.getLanguage().isEmpty() && isMultilingual) {
+                        fieldToAdd = field + "_" + entity.getLanguage();
+                    } else {
+                        fieldToAdd = field;
+                    }
+
+                    fieldValueToAdd = (String) entity.get(field);
+                    builder = builder.field(fieldToAdd, Arrays.asList(fieldValueToAdd));
+                    break;
+                } else {
+                    //We are making sure that the field is at-least added to the index.
+                    //This will help us tremendously when we are appending the data later in an already indexed document.
+                    builder = builder.field(field, Collections.emptyList());
+                }
+
+            }
+
+            builder = builder.endObject();
+            IndexRequest indexRequest = new IndexRequest(INDEX_NAME, docDef.getEntityField(), entity.getId())
+                    .source(builder);
+
+            /**
+             * We are creating an upsert request here instead of a simple insert request.
+             * The reason is we want to add a document if it does not exist with the given Subject Id (URI).
+             * But if the document exists with the same Subject Id, we want to do an update to it instead of deleting it and
+             * then creating it with only the latest field values.
+             * This functionality is called Upsert functionality and more can be learned about it here:
+             * https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html#upserts
+             */
+
+            //First Search of the field exists or not
+            SearchResponse existsResponse = client.prepareSearch(INDEX_NAME)
+                    .setTypes(docDef.getEntityField())
+                    .setQuery(QueryBuilders.existsQuery(fieldToAdd))
+                    .get();
+            String script;
+            if(existsResponse != null && existsResponse.getHits() != null && existsResponse.getHits().totalHits() > 0) {
```

Comment:
Updates now consist of two ES operations: first checking whether the entity+field exists, and then doing either an add or an update depending on the result. I wonder if a race condition is possible here, if many additions happen around the same time from multiple threads? I think that a single-operation atomic update would be preferable, if it is possible to implement with ES scripting. It would likely also perform much better, since the overhead of an extra ES HTTP request is likely quite significant considering that we are dealing with individual triples here.